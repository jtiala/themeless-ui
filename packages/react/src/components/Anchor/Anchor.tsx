import { CommonComponentProps, cn } from "@themeless-ui/utils";
import { AnchorHTMLAttributes, ReactNode } from "react";

type HTMLAnchorProps = AnchorHTMLAttributes<HTMLAnchorElement>;

export type AnchorProps = {
  /**
   * The URL that the hyperlink points to.
   */
  href?: HTMLAnchorProps["href"];

  /**
   * The relationship of the linked URL.
   *
   * If `rel` is not defined and the anchor has `target` as `_blank`, the following relationships will be added automatically:
   * - `noopener` to prevent reverse tabnabbing
   * - `noreferrer` to prevent passing referrer information to the target
   */
  rel?: HTMLAnchorProps["rel"];

  /**
   * Where to display the linked URL.
   */
  target?: AnchorHTMLAttributes<HTMLAnchorElement>["target"];

  /**
   * Causes the browser to treat the linked URL as a download.
   */
  download?: AnchorHTMLAttributes<HTMLAnchorElement>["download"];

  /**
   * A function that is called when the anchor is clicked.
   */
  onClick?: AnchorHTMLAttributes<HTMLAnchorElement>["onClick"];

  /**
   * A function that is called when the anchor is focused.
   */
  onFocus?: AnchorHTMLAttributes<HTMLAnchorElement>["onFocus"];

  /**
   * A function that is called when the anchor hast lost focus.
   */
  onBlur?: AnchorHTMLAttributes<HTMLAnchorElement>["onBlur"];

  /**
   * A function that is called when mouse enters the anchor.
   */
  onMouseEnter?: AnchorHTMLAttributes<HTMLAnchorElement>["onMouseEnter"];

  /**
   * A function that is called when mouse leaves the anchor.
   */
  onMouseLeave?: AnchorHTMLAttributes<HTMLAnchorElement>["onMouseLeave"];

  /**
   * A function that is called when one or more touch points are placed on the touch surface.
   */
  onTouchStart?: AnchorHTMLAttributes<HTMLAnchorElement>["onTouchStart"];

  /**
   * A function that is called when one or more touch points are removed from the touch surface.
   */
  onTouchEnd?: AnchorHTMLAttributes<HTMLAnchorElement>["onTouchEnd"];

  /**
   * Anchor content.
   */
  children?: ReactNode;
} & CommonComponentProps;

const className = cn("anchor");

/**
 * Display a hyperlink.
 */
export function Anchor({
  href,
  rel,
  target,
  download,
  onClick,
  onFocus,
  onBlur,
  onMouseEnter,
  onMouseLeave,
  onTouchStart,
  onTouchEnd,
  children,
  testId,
}: AnchorProps) {
  const openInNewTab = target === "_blank";
  const linkRelationships =
    rel || openInNewTab ? "noopener noreferrer" : undefined;

  return (
    <a
      href={href}
      rel={linkRelationships}
      target={target}
      download={download}
      onClick={onClick}
      onFocus={onFocus}
      onBlur={onBlur}
      onMouseEnter={onMouseEnter}
      onMouseLeave={onMouseLeave}
      onTouchStart={onTouchStart}
      onTouchEnd={onTouchEnd}
      className={className}
      data-testid={testId}
    >
      {children}
    </a>
  );
}
