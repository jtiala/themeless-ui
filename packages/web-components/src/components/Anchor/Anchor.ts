import style from "@themeless-ui/style/dist/anchor.css?inline";
import { cn } from "@themeless-ui/utils";
import { html, nothing, unsafeCSS } from "lit";
import { customElement, property } from "lit/decorators.js";
import {
  TUIBlurEvent,
  TUIClickEvent,
  TUIFocusEvent,
  TUIMouseEnterEvent,
  TUIMouseLeaveEvent,
  TUITouchEndEvent,
  TUITouchStartEvent,
} from "../../events";
import { TUIComponent } from "../TUIComponent";

const className = cn("anchor");

@customElement("tui-anchor")
export class Anchor extends TUIComponent {
  /**
   * The URL that the hyperlink points to.
   */
  @property()
  href?: HTMLAnchorElement["href"];

  /**
   * The relationship of the linked URL.
   *
   * If `rel` is not defined and the anchor has `target` as `_blank`, the following relationships will be added automatically:
   * - `noopener` to prevent reverse tabnabbing
   * - `noreferrer` to prevent passing referrer information to the target
   */
  @property()
  rel?: HTMLAnchorElement["rel"];

  /**
   * Where to display the linked URL.
   */
  @property()
  target?: HTMLAnchorElement["target"];

  /**
   * Causes the browser to treat the linked URL as a download.
   */
  @property()
  download?: HTMLAnchorElement["download"];

  static styles = unsafeCSS(style);

  private _handleClick() {
    this.dispatchEvent(new TUIClickEvent());
  }

  private _handleFocus() {
    this.dispatchEvent(new TUIFocusEvent());
  }
  private _handleBlur() {
    this.dispatchEvent(new TUIBlurEvent());
  }

  private _handleMouseEnter() {
    this.dispatchEvent(new TUIMouseEnterEvent());
  }

  private _handleMouseLeave() {
    this.dispatchEvent(new TUIMouseLeaveEvent());
  }

  private _handleTouchStart() {
    this.dispatchEvent(new TUITouchStartEvent());
  }

  private _handleTouchEnd() {
    this.dispatchEvent(new TUITouchEndEvent());
  }

  override render() {
    const openInNewTab = this.target === "_blank";
    const linkRelationships =
      this.rel || openInNewTab ? "noopener noreferrer" : undefined;

    return html`
      <a
        href="${this.href || nothing}"
        rel="${linkRelationships || nothing}"
        target="${this.target || nothing}"
        download="${this.download || nothing}"
        @click="${this._handleClick}"
        @focus="${this._handleFocus}"
        @blur="${this._handleBlur}"
        @mouseEnter="${this._handleMouseEnter}"
        @mouseLeave="${this._handleMouseLeave}"
        @touchStart="${this._handleTouchStart}"
        @touchEnd="${this._handleTouchEnd}"
        class="${className}"
        id="${this.id || nothing}"
        data-testid="${this.testId || nothing}"
      >
        <slot></slot>
      </a>
    `;
  }
}

declare global {
  interface HTMLElementTagNameMap {
    "tui-anchor": Anchor;
  }
}
