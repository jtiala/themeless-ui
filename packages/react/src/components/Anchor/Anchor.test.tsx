import {
  componentToJson,
  describe,
  expect,
  it,
  render,
  renderer,
  screen,
} from "../../../test";
import { Anchor } from "./Anchor";

const href = "https://example.com";
const children = "Example.com";

const defaultAnchor = (
  <Anchor testId="anchor" href={href}>
    {children}
  </Anchor>
);

const externalAnchor = (
  <Anchor testId="anchor" href={href} target="_blank">
    {children}
  </Anchor>
);

describe("Anchor", async () => {
  it("should render default anchor", () => {
    expect(componentToJson(renderer.create(defaultAnchor))).toMatchSnapshot();

    render(defaultAnchor);

    const anchor = screen.getByTestId("anchor");
    const linkRelationships = anchor.getAttribute("rel");

    expect(anchor).toBeInTheDocument();
    expect(anchor).toHaveTextContent(children);
    expect(linkRelationships).toBeFalsy();
  });

  it("should render external anchor", () => {
    expect(componentToJson(renderer.create(externalAnchor))).toMatchSnapshot();

    render(externalAnchor);

    const anchor = screen.getByTestId("anchor");
    const linkRelationships = (anchor.getAttribute("rel") || "").split(" ");

    expect(anchor).toBeInTheDocument();
    expect(anchor).toHaveTextContent(children);
    expect(anchor).toHaveAttribute("target", "_blank");
    expect(linkRelationships).toContain("noopener");
    expect(linkRelationships).toContain("noreferrer");
  });
});
