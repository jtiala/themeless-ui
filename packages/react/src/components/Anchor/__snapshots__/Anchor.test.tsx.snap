// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Anchor > should render default anchor 1`] = `
<a
  className="tui__anchor"
  data-testid="anchor"
  href="https://example.com"
>
  Example.com
</a>
`;

exports[`Anchor > should render external anchor 1`] = `
<a
  className="tui__anchor"
  data-testid="anchor"
  href="https://example.com"
  rel="noopener noreferrer"
  target="_blank"
>
  Example.com
</a>
`;
